test/e2e/storage/testsuites/base.go

```diff
@@ -465,3 +480,157 @@ func StartPodLogs(f *framework.Framework) func() {
+	metricsGrabber, err := metrics.NewMetricsGrabber(c, nil, true, false, true, false, false)
+
+	if err != nil {
+		framework.Failf("Error creating metrics grabber : %v", err)
+	}
+
+	if !metricsGrabber.HasRegisteredMaster() {
+		framework.Skipf("Environment does not support getting controller-manager metrics - skipping")
+	}
+
+	controllerMetrics, err := metricsGrabber.GrabFromControllerManager()
+	framework.ExpectNoError(err, "Error getting c-m metrics : %v", err)
+	totOps := getVolumeOpsFromMetricsForPlugin(controllerMetrics, pluginName)
+
+	nodes, err := c.CoreV1().Nodes().List(metav1.ListOptions{})
+	framework.ExpectNoError(err, "Error listing nodes: %v", err)
+	for _, node := range nodes.Items {
+		nodeMetrics, err := metricsGrabber.GrabFromKubelet(node.GetName())
+		framework.ExpectNoError(err, "Error getting Kubelet %v metrics: %v", node.GetName(), err)
+		totOps = addOpCounts(totOps, getVolumeOpsFromKubeletMetricsForPlugin(nodeMetrics, pluginName))
+	}
+
+	return totOps
+}
+
+func addOpCounts(o1 opCounts, o2 opCounts) opCounts {
+	totOps := opCounts{}
+	seen := sets.NewString()
+	for op, count := range o1 {
+		seen.Insert(op)
+		totOps[op] = totOps[op] + count
+		totOps[op] = totOps[op] + o2[op]
+	}
+	for op, count := range o2 {
+		if !seen.Has(op) {
+			totOps[op] = totOps[op] + count
+		}
+	}
+	return totOps
+}
+
+func getMigrationVolumeOpCounts(cs clientset.Interface, pluginName string) (opCounts, opCounts) {
+	if len(pluginName) > 0 {
+		var migratedOps opCounts
+		csiName, err := csilib.GetCSINameFromInTreeName(pluginName)
+		if err != nil {
+			framework.Logf("Could not find CSI Name for in-tree plugin %v", pluginName)
+			migratedOps = opCounts{}
+		} else {
+			csiName = "kubernetes.io/csi:" + csiName
+			migratedOps = getVolumeOpCounts(cs, csiName)
+		}
+		return getVolumeOpCounts(cs, pluginName), migratedOps
+	} else {
+		// Not an in-tree driver
+		framework.Logf("Test running for native CSI Driver, not checking metrics")
+		return opCounts{}, opCounts{}
+	}
+}
+
+func getTotOps(ops opCounts) int64 {
+	var tot int64 = 0
+	for _, count := range ops {
+		tot += count
+	}
+	return tot
+}
+
+func validateMigrationVolumeOpCounts(cs clientset.Interface, pluginName string, oldInTreeOps, oldMigratedOps opCounts) {
+	if len(pluginName) == 0 {
+		// This is a native CSI Driver and we don't check ops
+		return
+	}
+
+	newInTreeOps, newMigratedOps := getMigrationVolumeOpCounts(cs, pluginName)
+
+	if sets.NewString(strings.Split(*migratedPlugins, ",")...).Has(pluginName) {
+		// If this plugin is migrated based on the test flag storage.migratedPlugins
+		for op, count := range newInTreeOps {
+			if count != oldInTreeOps[op] {
+				framework.Failf("In-tree plugin %v migrated to CSI Driver, however found %v %v metrics for in-tree plugin", pluginName, count-oldInTreeOps[op], op)
+			}
+		}
+		totMigrated := getTotOps(newMigratedOps)
+		oldTotMigrated := getTotOps(oldMigratedOps)
+		if totMigrated-oldTotMigrated <= 0 {
```

Comment:
Some negative test cases may not cause the operation metrics to be incremented

Reply:
this is a good point. We could loosen the guarantees and only check:
if migrated -> no in-tree operations are done
if not migrated -> no migrated CSI Driver ops are done (we currently can't do this unless we add a "migrated" dimension to the metrics).

For the migrated case we at least know:
1) No in-tree operations were done in any test
2) The test passes so the e2e things we needed to do were achieved
these seem sufficient to determine that migration is "working". To determine that we're not "accidentally" using migration it is suffiecient to run all the tests with migration off without the driver installed.

So maybe this is enough. lmk what you think

Reply:
Yeah I think this is ok.